todo_app/app.py

```diff
@@ -1,54 +1,139 @@
-import os
-
+from os import access
+from functools import wraps
+import requests
 from flask import (
     Flask,
     render_template, 
     request, 
     redirect,
     url_for,
 )
+from flask_login import (
+    LoginManager, 
+    login_required, 
+    login_user, 
+    UserMixin, 
+    current_user,
+)
+from oauthlib.oauth2 import WebApplicationClient
 
 from todo_app.config import Config
 from todo_app.data.mongodb import MongoDB, CardStatus
 from todo_app.views.views import ViewModel
 
 
+class User(UserMixin):
+    def __init__(self, name):
+        self.id = name 
+
+
+login_manager = LoginManager()
+
+# Bad practice, but hardcoded for training purposes
+WRITER_LIST = ['Sm4o']
+
+
+@login_manager.unauthorized_handler
+def unauthenticated():
+    client = WebApplicationClient(Config().GITHUB_CLIENT_ID)
+    redirect_url = client.prepare_request_uri('https://github.com/login/oauth/authorize', 
+                                              redirect_uri='http://localhost:5000/login/callback')
+    return redirect(redirect_url)
+
+
+@login_manager.user_loader
+def load_user(name):
+    return User(name)
+
+
+def check_writer_role(func):
+    @wraps(func)
+    def decorated_view(*args, **kwargs):
+        if current_user.is_authenticated:
+            if current_user.id in WRITER_LIST:
+                return func(*args, **kwargs)
+            else:
+                return login_manager.unauthorized()
+        else:
+            # Only happens in tests or if no login_required decorator found
```

Comment:
If it's marked as writer_only and no login_required is present, should we let them through?
I'd be inclined to explicitly verify that LOGIN_DISABLED is on rather than just letting those cases slip through, though I agree in practice I think this is sound as it currently stands

Reply:
This was added specifically for E2E tests because if the user isn't logged in the `check_writer_role` decorator would give an error as `current_user` doesn't have attribute `id`. How can I check if LOGIN_DISABLED is True from within that decorator?

Reply:
I get the error when trying to create an item for example 
``` bash
AttributeError: 'AnonymousUserMixin' object has no attribute 'id'
```

Reply:
Got it, I can use `login_manager._login_disabled`